lib/ukdebug/README.md

```diff
@@ -0,0 +1,201 @@
+`ukdebug` menu in kconfig. Let's assume we are using x86. A possible QEMU
+command to start a VM that the GDB host can connect to looks like this:
+
+```bash
+qemu-system-x86_64 -kernel build/app-helloworld_qemu-x86_64 \
+                   -cpu max \
+                   -display none \
+                   -serial tcp::1235,server,nowait \
+                   -append "build/app-helloworld_qemu-x86_64 debug.gdbcon=0 --"
+```
+
+The only two interesting options here are `-serial` and `-append`.
+`-serial tcp::1235,server,nowait` makes QEMU create a TCP socket on
+`localhost:1235` that's connected to COM1 inside the VM. This TCP socket will
+be used to connect the GDB host to the serial console in the VM. `-append`
+specifies the kernel command line that's passed to the kernel at runtime.
+Passing the name of the image as the first parameter is an Unikraft convention.
+If the first parameter is a kernel parameter, this kernel parameter will be
+dropped, so be careful. With `debug.gdbcon=0`, we tell the GBD stub to use
+`con0` as the backing console. On x86 this is COM1 and it's connected to
+`localhost:1235`.
+
+You can use `-serial stdio` instead of `-serial tcp::1235,server,nowait` to
+see if the GDB stub is enabled and set up correctly. If you can run above
+command with `-serial stdio`, you should see some output that ends with
+something like the below (assuming you have debug messages enabled):
+
+```plain
+[    0.104746] Info: [libukdebug] <gdbstub.c @ 1237> Waiting for debugger connection on con0...
+[    0.104944] Info: [libukconsole] <console.c @  176> Registered con1: GDB virtual console, flags: --
+$S05#b8
+```
+
+The last line is the first packet that the GDB stub sends to the GDB host.
+Switch back to `-serial tcp::1235,server,nowait` and you'll stop seeing any
+output. Using `-serial stdio` can be useful when you're unsure if your
+configured kernel even boots.
+
+Note that the `-serial` flag is position-sensitive on x86. COM1 is connected
+to whatever the first occurrence (from left to right) of the `-serial` flag
+specifies. COM2 is configured based on the second occurrence of `-serial`, and
+so forth. E.g., as an alternative to the above, you could configure COM1 and
+COM2 in kconfig and then connect the GDB stub to COM2 and keep kernel output on
+COM1 (`debug.gdbcon` is `1` this time):
+
+```bash
+qemu-system-x86_64 -kernel build/app-helloworld_qemu-x86_64 \
+                   -cpu max \
+                   -display none \
+                   -serial stdio \
+                   -serial tcp::1235,server,nowait \
+                   -append "build/app-helloworld_qemu-x86_64 debug.gdbcon=1 --"
+```
+
+But let's stick to the initial example and assume that only COM1 is configured
+and that `debug.gdbcon=0`. Start the VM and in another terminal run GDB like
+this:
+
+```bash
+gdb build/app-helloworld_qemu-x86_64.dbg
+```
+
+There are two things to note here. First, don't forget the `.dbg` extension.
```

Comment:
> First, don't forget the `.dbg` extension. GDB won't find any symbols without it.

"Make sure you use the `.dgb` image, as it contains object symbols."

> Second, make sure you're using the right GDB binary. E.g., if you're developing on an x86 system and the kernel is compiled for Arm, you need to use GDB for Arm, not the stock GDB of your system.

I would avoid making specific assumptions about the host here, as with Mac it's also common to be on Arm and cross-compile for `x86_64`, so: "Second, make sure you're using a GDB binary compiled for the target architecture. If you're cross-compiling you must use the GDB binary provided by the cross-compiling toolchain".